pkg/psalabelsyncer/sccrolecache.go

```diff
@@ -0,0 +1,550 @@
+		Name:   saUserInfo.GetName(),
+		Groups: append(saUserInfo.GetGroups(), user.AllAuthenticated),
+		UID:    saUserInfo.GetUID(),
+		Extra:  saUserInfo.GetExtra(),
+	}
+
+	bindingObjs, err := getIndexedRolebindings(c.roleBindingIndexer, realSAUserInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterBindingObjs, err := getIndexedRolebindings(c.clusterRoleBindingIndexer, realSAUserInfo)
+	if err != nil {
+		return nil, err
+	}
+	objs := append(bindingObjs, clusterBindingObjs...)
+
+	sccs, err := c.sccLister.List(labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+
+	allowedSCCs := sets.NewString()
+	for _, scc := range sccs {
+		if sccAllowsSA(scc, realSAUserInfo) {
+			allowedSCCs.Insert(scc.Name)
+		}
+	}
+
+	for _, o := range objs {
+		rb, err := newRoleBindingObj(o)
+		if err != nil {
+			// this would be rather weird
+			return nil, err
+		}
+
+		roleCachedKey := fmt.Sprintf("/%s", rb.RoleRef().Name)
+		if rb.RoleRef().Kind == "Role" {
+			roleCachedKey = rb.Namespace() + roleCachedKey
+		}
+
+		c.usefulRolesLock.Lock()
+		// this role does not have SCC-related rules
+		cachedAllowedSCCs := c.usefulRoles[roleCachedKey]
+		c.usefulRolesLock.Unlock()
+		if len(cachedAllowedSCCs) == 0 {
+			continue
+		}
+
+		// we particularly care only about Roles in the SA NS
+		if roleRef := rb.RoleRef(); roleBindingAppliesToNS(rb, serviceAccount.Namespace) && roleRef.APIGroup == rbacv1.GroupName {
+			allowedSCCs.Insert(cachedAllowedSCCs...)
+		}
+	}
+
+	return allowedSCCs, nil
+}
+
+// getRoleFromRoleRef tries to retrieve the role or clusterrole from roleRef.
+func (c *SAToSCCCache) getRoleFromRoleRef(ns string, roleRef rbacv1.RoleRef) (RoleInterface, error) {
+	var err error
+	var role interface{}
+	switch kind := roleRef.Kind; kind {
+	case "Role":
+		role, err = c.roleLister.Roles(ns).Get(roleRef.Name)
+		if err != nil {
+			return nil, err
+		}
+
+	case "ClusterRole":
+		role, err = c.clusterRoleLister.Get(roleRef.Name)
+		if err != nil {
+			return nil, err
+		}
+
+	default:
+		return nil, fmt.Errorf("unknown kind in roleRef: %s", kind)
+	}
+
+	return NewRoleObj(role)
+}
+
+// IsRoleBindingRelevant returns true if the cluster/rolebinding supplied binds
+// to a Role that provides access to at least one of the SCCs available in the
+// cluster.
+func (c *SAToSCCCache) IsRoleBindingRelevant(obj interface{}) bool {
+	rb, err := newRoleBindingObj(obj)
+	if err != nil {
+		klog.Warningf("unexpected error, this may be a bug: %v", err)
+		return false
+	}
+
+	role, err := c.getRoleFromRoleRef(rb.Namespace(), rb.RoleRef())
+	if err != nil {
+		klog.Infof("failed to retrieve a role for a rolebinding ref: %v", err)
+		return false
+	}
+
+	c.usefulRolesLock.Lock()
+	defer c.usefulRolesLock.Unlock()
+
+	return len(c.usefulRoles[usefulRolesKey(role)]) != 0
+}
+
+func (c *SAToSCCCache) syncRoleCacheSingleSCCAdded(role RoleInterface, rules []rbacv1.PolicyRule, scc *securityv1.SecurityContextConstraints) {
+	if c.usefulRolesLock.TryLock() {
```

Comment:
I am sceptical about this. Are you worried about deadlocks? If so we should rather proof they can occur and prevent them. As per https://pkg.go.dev/sync#Mutex.TryLock:

> Note that while correct uses of TryLock do exist, they are rare, and use of TryLock is often a sign of a deeper problem in a particular use of mutexes.

Reply:
This is not about deadlocks. It is a preventive measure for the programmer using the function without a lock.

Reply:
as discussed OOB this is not enough of a justification, xref https://github.com/golang/go/issues/45435#issuecomment-832910477 and https://github.com/golang/go/issues/45435#issuecomment-841028093